### ISSUE TITLE
Junit report numbers sub-tests regardless of numbers in TAP report

### ISSUE THREAD

**Issue by ppaalanen** [mesonbuild/meson#8350]
I'm working on https://gitlab.freedesktop.org/wayland/weston/-/merge_requests/557 where one patch is deliberately stopping to print the test counter number in the TAP report. My intention was that not printing the counter would make the number disappear from the Gitlab test reports. However, the number is still there, as can be seen in that MR.

Example from Junit output:
```
<testcase classname="weston.alpha-blending" name="2 GL alpha_blend_monotonic/1">
```

The number 2 should not be there. I do not print it, but Meson 0.57.0 adds it. For reference, this is the TAP output for the whole test program:
```
1..2
ok pixman alpha_blend_monotonic/1
ok GL alpha_blend_monotonic/1
```

Should Meson not add that number?

**Comment by bonzini** [comment]
TAP tests are numbered implicitly (you can see that from the fact that you have the `1..2` test plan). While you don't _have to_ specify them in the TAP file, I suggest that you do in case one day you have a testcase whose first word is a number, like

```
1..1
ok 256 colors
```

**Comment by ppaalanen** [comment]
Right, but why does Meson print the numbers into Junit? That should be unrelated to whether TAP needs/has numbers or not.

**Comment by ppaalanen** [comment]
I have reinstated the case numbers in my TAP report, and I would like Meson to drop those numbers when converting TAP to Junit.

**Comment by bonzini** [comment]
> Right, but why does Meson print the numbers into Junit?

I guess it's a matter of taste. I like having numbers because if you see a failure on test N you can put a watchpoint on the counter becoming N (or N-1 depending on the harness). You don't have to think too much about what function corresponds to which testcase.

I admit that Meson is not consistent though. The console log does not have the numbers, which however wasn't something I really liked. I only did it that way because the test number precedes the subtest (to make sense of tests running in parallel).

    ▶  6/10 skipped subtest        SKIP

On the other hand, the text log has the full TAP output and doesn't mix parallel tests, so it does not have the same issue and numbers are there:

```
 6/10 some test                      OK             22.71s   1 subtests passed
12:05:09 MALLOC_PERTURB_=37 '/home/pbonzini/f/exec'
----------------------------------- output -----------------------------------
ok 1 passed subtest
ok 2 skipped subtest # skip
------------------------------------------------------------------------------
```

In my opinion JUnit was more like the latter and so I left the numbers in.

**Comment by ppaalanen** [comment]
In general, a Weston test program (which there are several, each a Meson `test()`) has M "fixture setups", K test functions, and each function can have any number of sub-cases reported independently in TAP. For each fixture setup, it runs each function, and all sub-cases inside the function, by default.

You can pick and choose which cases you run: choose a fixture setup and run everything under it; choose a function and run it on all fixture setups; choose a function and a specific sub-case by number, run it alone on a specific fixture or on all fixture setups.

In Weston's test suite, the TAP numbers are artificial, generated on the fly, and do not have a fixed correspondence to the actual tests. You can tell a test program to run a subset of the tests it contains and whatever happens to run gets numbered from 1 to N.

So the TAP numbering is completely useless for Weston. Granted though, Gitlab CI always runs everything, so the TAP numbers are stable as long as no sub-case is added or removed. The test suite just doesn't use TAP numbers for anything as you can pick your "row" by choosing the fixture setup and/or function and maybe sub-case number.

I have no way of omitting the useless numbers. OTOH, if the numbers are useful to you, you could include them in your test names explicitly.

Do we need something like a `protocol: tap-drop-numbers`?

Or just live with this?

**Comment by bonzini** [comment]
I think we need to get other opinions. @eli-schwartz, you were looking at Meson TAP support in the past, so what do you think?

**Comment by eli-schwartz** [comment]
The pacman tap runner already reports e.g.

```
1..36
ok 1 - compose-multiple-template-dirs exitcode
ok 2 - compose-multiple-template-dirs output
ok 3 - compose-multiple-template-dirs resulting PKGBUILD
ok 4 - invalid-key exitcode
ok 5 - invalid-key output
[...]
```

I don't really see a reason for meson to add a second copy of the number, but on the other hand, testlog.junit.xml does not seem to do so for me (I have never looked at this file before so I'm not positive I know what I'm looking at).

**Comment by bonzini** [comment]
It's not a second copy; the question is whether to remove the only instance of the number (in 0.57, the test names would be like `5 - invalid-key output`).

**Comment by eli-schwartz** [comment]
> I have reinstated the case numbers in my TAP report, and I would like Meson to drop those numbers when converting TAP to Junit.

??? So does meson already drop the numbers?

**Comment by bonzini** [comment]
> So does meson already drop the numbers?

No, it doesn't drop them and it adds the implicit numbering if they are left out by the test.

The main advantage of having numbers are that they allow disambiguating tests with identical text (or without text). The main disadvantage is that they are unstable (they change even though the test remains the same).

**Comment by ppaalanen** [comment]
Right, it makes no difference in the Junit output of Meson at all whether my TAP report has be numbers or not. I did not remove the numbers from my TAP reports in the end because of https://github.com/mesonbuild/meson/issues/8350#issuecomment-780468532 .

Apparently Meson does recognize the TAP number if it is present, and adds the numbering if it is not present. This is according to the TAP specification, and it's fine.

I would like to omit the TAP numbers from the **Junit** test report, because on the Gitlab test result pages created from the Junit report that number is useless.

The TAP numbering is also not unique. Every test program (Meson's `test()` instance) starts the numbering from 1, and we have many test programs.

**Comment by bonzini** [comment]
> The TAP numbering is also not unique

The suite+number+test is unique though.

**Comment by ppaalanen** [comment]
> The (JUnit) suite+number+test is unique though.

Sure, but then the number is unnecessary.

**Comment by ppaalanen** [comment]
This is the Gitlab test report:
![Gitlab-Junit-test-report](https://user-images.githubusercontent.com/2331435/108346896-90c34600-71e8-11eb-99c5-1d326b0c38db.png)

There is no use for the leading number in the Name column. It is not unique alone, it is not necessary for uniquely identifying a row, it is not stable when tests are changed, and it can only confuse the reader as it is not used at all in any developer facing interfaces.

**Comment by bonzini** [comment]
Well, sort of.  suite+test is unique in your case, but not in general. suite+number is unique but unreadable. suite+number+test is unique and readable and unstable. I can see why you don't want it, and I don't like the instability of the number, but I'm not sure removing it is a good idea. You can probably use a quick "sed" to remove it in the CI run, if you want.
